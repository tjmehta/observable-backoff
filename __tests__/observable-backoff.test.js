/* eslint-env jest */
const Observable = require('rxjs/Observable').Observable
const Subscription = require('rxjs/Subscription').Subscription
const StaticObservable = require('static-observable')
const timeout = require('timeout-then')
require('observable-backoff')

describe('observable-backoff', () => {
  describe('no errors', () => {
    it('should work like a normal observable', () => {
      const observable = new StaticObservable()
      const callbacks = {
        onNext: jest.fn(),
        onError: jest.fn(),
        onComplete: jest.fn()
      }
      observable
        .backoff()
        .subscribe(
          callbacks.onNext,
          callbacks.onError,
          callbacks.onComplete
        )
      observable.next(1)
      expect(callbacks).toMatchSnapshot()
      observable.next(2)
      expect(callbacks).toMatchSnapshot()
      observable.next(3)
      expect(callbacks).toMatchSnapshot()
      observable.complete()
      expect(callbacks).toMatchSnapshot()
    })
  })

  describe('errors', () => {
    it('should exponentially backoff and retry', () => {
      const mocks = {
        onNext: jest.fn(),
        onError: jest.fn(),
        onComplete: jest.fn(),
        subscribe: jest.fn(),
        optsOnError: jest.fn()
      }
      let failCount = 0
      const subscribe = (observer) => {
        mocks.subscribe(observer.constructor.name) // for snapshot
        observer.error(failCount)
        failCount++
        return new Subscription()
      }
      const observable = new Observable(subscribe)
      const opts = {
        // 5, 10, 20
        minTimeout: 5,
        maxTimeout: 10 * Math.pow(2, 2),
        factor: 2,
        retries: 2,
        onError: mocks.optsOnError
      }
      observable
        .backoff(opts)
        .subscribe(
          mocks.onNext,
          mocks.onError,
          mocks.onComplete
        )
      return timeout(0)
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(5 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(10 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(20 + 1)) // extra to ensure it doesn't retry too many times
        .then(() => expect(mocks).toMatchSnapshot())
    })
    it('should exponentially backoff and stop if onError returns true', () => {
      const mocks = {
        onNext: jest.fn(),
        onError: jest.fn(),
        onComplete: jest.fn(),
        subscribe: jest.fn(),
        optsOnError: jest.fn(function (err, final, count) {
          const dontRetry = count === 3
          return dontRetry
        })
      }
      let failCount = 0
      const subscribe = (observer) => {
        mocks.subscribe(observer.constructor.name) // for snapshot
        observer.error(failCount)
        failCount++
        return new Subscription()
      }
      const observable = new Observable(subscribe)
      const opts = {
        // 5, 10, 20
        minTimeout: 5,
        maxTimeout: 10 * Math.pow(2, 2),
        factor: 2,
        retries: 2,
        onError: mocks.optsOnError
      }
      observable
        .backoff(opts)
        .subscribe(
          mocks.onNext,
          mocks.onError,
          mocks.onComplete
        )
      return timeout(0)
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(5 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(10 + 1)) // stopped retrying..
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(20 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
    })
  })
})
