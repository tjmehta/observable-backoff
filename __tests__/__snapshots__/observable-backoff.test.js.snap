// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`observable-backoff errors should exponentially backoff and retry 1`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff errors should exponentially backoff and retry 2`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
      Array [
        1,
        false,
        2,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff errors should exponentially backoff and retry 3`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction] {
    "calls": Array [
      Array [
        2,
      ],
    ],
  },
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
      Array [
        1,
        false,
        2,
      ],
      Array [
        2,
        true,
        3,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff errors should exponentially backoff and retry 4`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction] {
    "calls": Array [
      Array [
        2,
      ],
    ],
  },
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
      Array [
        1,
        false,
        2,
      ],
      Array [
        2,
        true,
        3,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff errors should exponentially backoff and stop if onError returns true 1`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff errors should exponentially backoff and stop if onError returns true 2`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
      Array [
        1,
        false,
        2,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff errors should exponentially backoff and stop if onError returns true 3`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction] {
    "calls": Array [
      Array [
        2,
      ],
    ],
  },
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
      Array [
        1,
        false,
        2,
      ],
      Array [
        2,
        true,
        3,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff errors should exponentially backoff and stop if onError returns true 4`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction] {
    "calls": Array [
      Array [
        2,
      ],
    ],
  },
  "onNext": [MockFunction],
  "optsOnError": [MockFunction] {
    "calls": Array [
      Array [
        0,
        false,
        1,
      ],
      Array [
        1,
        false,
        2,
      ],
      Array [
        2,
        true,
        3,
      ],
    ],
  },
  "subscribe": [MockFunction] {
    "calls": Array [
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
      Array [
        "RetryWhenSubscriber",
      ],
    ],
  },
}
`;

exports[`observable-backoff no errors should work like a normal observable 1`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction] {
    "calls": Array [
      Array [
        1,
      ],
    ],
  },
}
`;

exports[`observable-backoff no errors should work like a normal observable 2`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction] {
    "calls": Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
    ],
  },
}
`;

exports[`observable-backoff no errors should work like a normal observable 3`] = `
Object {
  "onComplete": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction] {
    "calls": Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
      Array [
        3,
      ],
    ],
  },
}
`;

exports[`observable-backoff no errors should work like a normal observable 4`] = `
Object {
  "onComplete": [MockFunction] {
    "calls": Array [
      Array [],
    ],
  },
  "onError": [MockFunction],
  "onNext": [MockFunction] {
    "calls": Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
      Array [
        3,
      ],
    ],
  },
}
`;
